File: packages/mobile/ios/App/App/OpenChamberSharePlugin.swift
import Capacitor

@objc(OpenChamberSharePlugin)
class OpenChamberSharePlugin: CAPPlugin, CAPBridgedPlugin {
    let identifier = "OpenChamberSharePlugin"
    let jsName = "OpenChamberShare"
    let pluginMethods: [CAPPluginMethod] = [CAPPluginMethod(name: "updateCatalog", returnType: CAPPluginReturnPromise), CAPPluginMethod(name: "donateAssistantInteraction", returnType: CAPPluginReturnPromise), CAPPluginMethod(name: "listPending", returnType: CAPPluginReturnPromise), CAPPluginMethod(name: "ack", returnType: CAPPluginReturnPromise), CAPPluginMethod(name: "releaseFiles", returnType: CAPPluginReturnPromise)]
    // Capacitor's getArray(_:type:) treats nested dictionaries poorly; cast via Any.
    @objc func updateCatalog(_ call: CAPPluginCall) {
        guard let raw = call.getArray("entries") else { call.reject("entries is required"); return }
        let entries = raw.compactMap { $0 as? [String: Any] }
        do { try OpenChamberShareStore.updateCatalog(entries); call.resolve() } catch { call.reject(error.localizedDescription) }
    }
    @objc func donateAssistantInteraction(_ call: CAPPluginCall) {
        guard let serverInstanceID = call.getString("serverInstanceID"),
              let assistantID = call.getString("assistantID"),
              let name = call.getString("name") else { call.reject("assistant identity is required"); return }
        let target: [String: Any] = ["serverInstanceID": serverInstanceID, "assistantID": assistantID, "name": name, "avatarSeed": call.getString("avatarSeed") ?? assistantID]
        OpenChamberShareStore.donateAssistantInteraction(target: target) { error in
            DispatchQueue.main.async {
                if let error { call.reject(error.localizedDescription) } else { call.resolve() }
            }
        }
    }
    @objc func listPending(_ call: CAPPluginCall) { do { let data = try JSONEncoder().encode(try OpenChamberShareStore.pending()); call.resolve(["envelopes": try JSONSerialization.jsonObject(with: data)]) } catch { call.reject(error.localizedDescription) } }
    @objc func ack(_ call: CAPPluginCall) { guard let operationID = call.getString("operationID") else { call.reject("operationID is required"); return }; do { try OpenChamberShareStore.acknowledge(operationID); call.resolve() } catch { call.reject(error.localizedDescription) } }
    @objc func releaseFiles(_ call: CAPPluginCall) { guard let operationID = call.getString("operationID") else { call.reject("operationID is required"); return }; do { try OpenChamberShareStore.release(operationID); call.resolve() } catch { call.reject(error.localizedDescription) } }
    func emitReceived(_ operationID: String) { notifyListeners("shareReceived", data: ["operationID": operationID]) }
}

File: packages/mobile/android/app/src/main/java/com/openchamber/app/OpenChamberSharePlugin.java
package com.openchamber.app;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.JSObject;
import com.getcapacitor.annotation.CapacitorPlugin;
import org.json.JSONArray;

@CapacitorPlugin(name = "OpenChamberShare")
public class OpenChamberSharePlugin extends Plugin {
    @PluginMethod public void updateCatalog(PluginCall call) { try { JSONArray entries = call.getArray("entries"); if (entries == null) throw new IllegalArgumentException("entries is required"); OpenChamberShareStore.updateCatalog(getContext(), entries); call.resolve(); } catch (Exception error) { call.reject(error.getMessage(), error); } }
    @PluginMethod public void listPending(PluginCall call) { try { JSObject result = new JSObject(); result.put("envelopes", OpenChamberShareStore.pending(getContext())); call.resolve(result); } catch (Exception error) { call.reject(error.getMessage(), error); } }
    @PluginMethod public void ack(PluginCall call) { String id = call.getString("operationID"); if (id == null || !id.matches("[A-Za-z0-9-]{1,80}")) { call.reject("A valid operationID is required."); return; } try { OpenChamberShareStore.acknowledge(getContext(), id); call.resolve(); } catch (Exception error) { call.reject(error.getMessage(), error); } }
    @PluginMethod public void releaseFiles(PluginCall call) { String id = call.getString("operationID"); if (id == null || !id.matches("[A-Za-z0-9-]{1,80}")) { call.reject("A valid operationID is required."); return; } try { OpenChamberShareStore.releaseAcknowledged(getContext(), id); call.resolve(); } catch (Exception error) { call.reject(error.getMessage(), error); } }
    void emitReceived(String operationID) { JSObject event = new JSObject(); event.put("operationID", operationID); notifyListeners("shareReceived", event); }
}
